Update Coding Guidelines (#441)
Azure/cosmos-explorer: CODING_GUIDLINES.md

# Coding Guidelines and Recommendations

The Stack:

- Language: [TypeScript](https://www.typescriptlang.org/)
- Framework: [React](https://reactjs.org/) + [Knockout](https://knockoutjs.com/)(Deprecated)
- Bundler: [Webpack](https://webpack.js.org/)
- Unit Test Runner: [Jest](https://jestjs.io/)
- Code Formatter: [Prettier](https://prettier.io/)
- Component Framework: [Fluent](https://developer.microsoft.com/en-us/fluentui#/)
- Linter: [ESLint](https://eslint.org/) + [TSLint](https://palantir.github.io/tslint/)(Deprecated)
- End to End Test Runner: [Jest](https://jestjs.io/)
- End to End Test Framework: [Puppeteer](https://developers.google.com/web/tools/puppeteer)
- Transpiler: [Babel](https://babeljs.io/) + [TypeScript](https://www.typescriptlang.org/)

## Requirements

> Cosmos Explorer has been under constant development for over 5 years. As a result there are many different patterns and practices in the code base. This document serves as a guide to how we _should_ be writing code and helps avoid propagating practices which are no longer preferred. Each requirement in this document is labelled and color-coded to show the relative importance. In order from highest importance to lowest importance:

Reviewer: ```suggestion
Cosmos Explorer has been under constant development for over 5 years. As a result there are many different patterns and practices in the code base. This document serves as a guide to how we write code and helps avoid propagating practices which are no longer preferred. Each requirement in this document is labelled and color-coded to show the relative importance. In order from highest importance to lowest importance:
```

You should just assert that this is how we write code. `MUST` vs `SHOULD` is defined by your emojis below, so you don't need to add "_should_" here. Requirements are best done assertively.

✅ DO this. If you feel you need an exception, engage with the project owners _prior_ to implementation.

⛔️ DO NOT do this. If you feel you need an exception, engage with the project owners _prior_ to implementation.

☑️ YOU SHOULD strongly consider this but it is not a requirement. If not following this advice, please comment code with why and proactively begin a discussion as part of the PR process.

⚠️ YOU SHOULD NOT strongly consider not doing this. If not following this advice, please comment code with why and proactively begin a discussion as part of the PR process.

> ✔️ YOU MAY consider this advice if appropriate to your situation. Other team members may comment on this as part of PR review, but there is no need to be proactive.

Reviewer: ✅ vs ✔ isn't very different. Maybe a 💭 for "think about this" or 🤔 "ponder"

```suggestion
💭 YOU MAY consider this advice if appropriate to your situation. Other team members may comment on this as part of PR review, but there is no need to be proactive.
```

## Development Environment

☑️ YOU SHOULD use VSCode and install the following extensions:

- [Prettier](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode)
- [ESLint](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint)

This setup will catch most linting/formatting/type errors as you develop.

> ☑️ YOU SHOULD use the [GitHub CLI](https://cli.github.com/). It has helpful workflows for submitting PRs as well as for checking out other team members PRs.

Reviewer 2: > use the [](start = 13, length = 8)

why should this be a should :)

Author: Changed

> ✔️ YOU MAY use Windows, Linux (including WSL), or OSX. We have team members on all three environments. When modifying any engineering systems please keep cross platform compatability in mind.

Reviewer: I'd break the second sentence into a `MUST`

⚠️ YOU SHOULD NOT disable any lint rules. They are there as guidance and to catch common mistakes. You will find places we disable specific lint rules, but is should be exceptional. If a rule does need to be disabled, prefer disabling a specific line instead of the entire file.

⛔️ DO NOT add any tslint rules. TSLint has been deprecated and is on track to be removed. Always prefer ESLint rules.

## UI Components

☑️ YOU SHOULD write new components using React. We are actively migrating Cosmos Explorer off of Knockout.

☑️ YOU SHOULD use Fluent UI components. These are highly accesible and composable. They allow us to build upon the work of the Fluent team and lead to a lower total cost of ownership for our UI code.

### React

⛔️ DO NOT use inheritance for sharing component behavior. The React documentation covers this topic in detail https://reactjs.org/docs/composition-vs-inheritance.html

☑️ YOU SHOULD use pure functional components when no state is required

✔️ YOU MAY consider using functional (hooks) or class components. The project contains examples of both and neither is strongly preferred right now.

## Libraries

> ⛔️ DO NOT add new libraries to package.json. Adding libraries has the possibility to bring in code that explodes our bundled size and/or attempts to run NodeJS code in the browser. Consult with project owners first.

Reviewer: This feels like a `SHOULD NOT` instead of a `MUST NOT`, given that the guidance is to consult with project owners first.

Reviewer: Or a `MUST` consult with the project owners before adding a package which is more specific.

## Testing

⛔️ DO NOT decrease test coverage. Unit/Functional test coverage is checked as part of the CI process.

### Unit Tests

✅ DO write unit tests for non-UI and utility code.

☑️ YOU SHOULD abstract non-UI and utility code so it can run either the NodeJS or Browser environment

### Functional(Component) Tests

✅ DO write tests for UI components

✅ DO use either Enzyme or React Testing Library to perform component tests.

### Mocking

> ☑️ YOU SHOULD try to write code that does not require mocking

Reviewer: ```suggestion
☑️ YOU SHOULD write code that does not require mocking
```

"should try" is passive language

⚠️ YOU SHOULD NOT strongly consider not doing this. If not following this advice, please comment code with why and proactively begin a discussion as part of the PR process.

💭 YOU MAY consider this advice if appropriate to your situation. Other team members may comment on this as part of PR review, but there is no need to be proactive.

## Development Environment

☑️ YOU SHOULD use VSCode and install the following extensions:

- [Prettier](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode)
- [ESLint](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint)

This setup will catch most linting/formatting/type errors as you develop.

☑️ YOU SHOULD use the [GitHub CLI](https://cli.github.com/). It has helpful workflows for submitting PRs as well as for checking out other team members PRs.

💭 YOU MAY use Windows, Linux (including WSL), or OSX. We have team members on all three environments.

✅ DO maintain cross platform compatability when modifying any engineering systems.

## Formatting

> ✅ DO use [Prettier](https://prettier.io/) to format your code. This will be preformed automatically if using the recommended editor setup. Otherwise `npm run format` will also format code.

Reviewer 2: >DO use [Prettier [](start = 2, length = 16)

how about we have this close to where you talk about prettier earlier in line 34. that will give more context.

Author: This is a good point. Maybe I can move everything in the "Stack" section down along side the todos

## Linting

> ✅ DO use [ESLint](https://eslint.org/) to check for linting errors your code. This can also be performed by `npm run lint`

Reviewer 2: >O use [ESLin [](start = 3, length = 12)

same here ..move up?

💭 YOU MAY consider adding new lint rules. If you find yourself performing "nits" as part of PR review, consider if it is possible to add a lint rule that will autuomatically catch the error in the future

⚠️ YOU SHOULD NOT disable any lint rules. They are there as guidance and to catch common mistakes. You will find places we disable specific lint rules, but is should be exceptional. If a rule does need to be disabled, prefer disabling a specific line instead of the entire file.

⛔️ DO NOT add any tslint rules. TSLint has been deprecated and is on track to be removed. Always prefer ESLint rules.

> 

Reviewer 2: Can we group them by category? All do this together, all do nots together , etc for better readability

Author: How about grouping the categories inside each component? IE all the `DOs` go together under linting, then under UI, etc. I think keeping a global `DOs` list will be worse for people finding stuff

## Development Environment

☑️ YOU SHOULD

- Use VSCode and install the following extensions. This setup will catch most linting/formatting/type errors as you develop:
  - [Prettier](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode)
  - [ESLint](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint)

💭 YOU MAY

- Use the [GitHub CLI](https://cli.github.com/). It has helpful workflows for submitting PRs as well as for checking out other team members PRs.
- Use Windows, Linux (including WSL), or OSX. We have team members developing on all three environments.

✅ DO

- Maintain cross platform compatability when modifying any engineering systems.

## Code Formatting

✅ DO

- Use [Prettier](https://prettier.io/) to format your code
  - This will be preformed automatically if using the recommended editor setup.
  - `npm run format` will also format code.

## Linting

✅ DO

- Use [ESLint](https://eslint.org/) to check for linting errors your code.
  - This will be preformed automatically if using the recommended editor setup.

>   - `npm run lint` was also format your code

Reviewer 2: >was [](start = 19, length = 3)

will

💭 YOU MAY

- Consider adding new lint rules.
  - If you find yourself performing "nits" as part of PR review, consider adding a lint rule that will autuomatically catch the error in the future

⚠️ YOU SHOULD NOT

- Disable lint rules
  - Lint rules exisst as guidance and to catch common mistakes.

>   - You will find places we disable specific lint rules, but is should be exceptional.

Reviewer 2: >s [](start = 62, length = 1)

it

  - `npm run lint` will also check for linting errors

💭 YOU MAY

- Consider adding new lint rules.
  - If you find yourself performing "nits" as part of PR review, consider adding a lint rule that will automatically catch the error in the future

⚠️ YOU SHOULD NOT

- Disable lint rules
  - Lint rules exist as guidance and to catch common mistakes
  - You will find places we disable specific lint rules however it should be exceptional.
  - If a rule does need to be disabled, prefer disabling a specific line instead of the entire file.

⛔️ DO NOT

- Add [TSLint](https://palantir.github.io/tslint/) rules
  - TSLint has been deprecated and is on track to be removed
  - Always prefer ESLint rules

## UI Components

☑️ YOU SHOULD

- Write new components using [React](https://reactjs.org/). We are actively migrating Cosmos Explorer off of [Knockout](https://knockoutjs.com/).
- Use [Fluent](https://developer.microsoft.com/en-us/fluentui#/) components.
  - Fluent components are designed to be highly accessible and composable
  - Using Fluent allows us to build upon the work of the Fluent team and leads to a lower total cost of ownership for UI code

> 

Reviewer 3: Naming convention: do not suffix with "Component". Even though the code has examples of it, we are ending the practice.

### React

☑️ YOU SHOULD

- Use pure functional components when no state is required

💭 YOU MAY

- Use functional (hooks) or class components
  - The project contains examples of both
  - Neither is strongly preferred at this time

⛔️ DO NOT

- Use inheritance for sharing component behavior.
  - React documentation covers this topic in detail https://reactjs.org/docs/composition-vs-inheritance.html

## Libraries

⚠️ YOU SHOULD NOT

- Add new libraries to package.json.
  - Adding libraries may bring in code that explodes the bundled size or attempts to run NodeJS code in the browser
  - Consult with project owners for help with library selection if one is needed

⛔️ DO NOT

> - Use underscore

Author: Agreed. -> underscore.js

### Functional(Component) Tests

✅ DO

- Write tests for UI components
- Write your tests using [Jest](https://jestjs.io/)
- Use either Enzyme or React Testing Library to perform component tests.

### Mocking

✅ DO

- Use Jest's built-in mocking helpers

☑️ YOU SHOULD

- Write code that does not require mocking
- Build components that do not require mocking extremely large or difficult to mock objects (like Explorer.ts). Pass _only_ what you need.

⛔️ DO NOT

- Use sinon.js for mocking
  - Sinon has been deprecated and planned for removal

### End to End Tests

✅ DO

- Use [Puppeteer](https://developers.google.com/web/tools/puppeteer) and [Jest](https://jestjs.io/)
- Write or modify an existing E2E test that covers the primary use case of any major feature.
  - Use caution. Do not try to cover every case. End to End tests can be slow and brittle.

☑️ YOU SHOULD

- Write tests that use accessible attributes to perform actions. Role, Title, Label, etc
  - More information https://testing-library.com/docs/queries/about#priority

⚠️ YOU SHOULD NOT

- Add test specfic `data-*` attributes to dom elements

>   - This is a common current practice, but one we would like to avoid in the future

Reviewer 3: Can you elaborate with an example? I'm curious.

Author: I think this covers some of it? https://testing-library.com/docs/queries/about#priority

The gist is that we should prefer end to end test to interact with our app as an actual user to be truly end to end. Using data-test attributes works around that goal. Also they are not accessible. By making treating our end user needs accessibility we are also verifying that a11y support.

Author: Added some more info about this

> 

Reviewer 3: - Add arbitrary time delays to wait for page to render or element to be ready. All the time delays add up and slow down testing. Prefer using the framework's "wait for..." functionality.

### Migrating Knockout to React

✅ DO

- Consult other team members before beginning migration work. There is a significant amount of flux in patterns we are using and it is important we do not propagate incorrect patterns.
- Start by converting HTML to JSX: https://magic.reactjs.net/htmltojsx.htm. Add functionality as a second step.

☑️ YOU SHOULD

- Write React components that require no dependency on Knockout or observables to trigger rendering.

## Browser Support

✅ DO

- Support IE11
  - In practice, this should not need to be considered as part of a normal development workflow
  - Polyfills and transpilation are already provided by our engineering systems.

>   - This requirement will be removed on March 30th, 2020 when Azure drops IE11 support.

Reviewer 2: >2020 [](start = 52, length = 4)

2021 :)
